lib/rubocop/cop/discourse/plugins/check_plugin_is_enabled.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Discourse
+      module Plugins
+        # When adding methods to an existing class, add a guard clause to check
+        # whether the plugin is enabled.
```

Comment:
Why would we require this on every method? IMO it depends on the scenario. Adding a guard clause only makes sense in a small number of cases.

e.g. this would be fine:

```ruby
class ::Topic
  def new_method
    "some response"
  end
end
```

```ruby
add_to_serializer(:topic, :new_method)
```

Because the add_to_serializer ensures that `Topic#new_method` will only be called when the plugin is enabled.

Reply:
I don’t think your example is a good one as `add_to_serializer` doesn’t work like that :sweat_smile: (it needs a block defining the method body and `:topic` will actually search for `TopicSerializer`, not the `Topic` model)

This cop only checks for new methods added to existing classes inside `plugin.rb`. If you’re adding stuff to a serializer, then using `add_to_serializer` will take care of the offenses. For other cases, either add a check, so the method won’t do anything if the plugin is disabled or extract the new code to a dedicated mixin. One case we wanted to avoid was a new method defined on a model which was then called by a callback in that same model, but there wasn’t a guard clause. So that method is called every time even if the plugin is disabled.

If we could wrap automatically new methods defined by a plugin so we check whether the plugin is disabled, it would be nice. Not the same kind of work though :sweat_smile:

Reply:
> add_to_serializer doesn’t work like that 

🤦 ah yup, my bad. So the example would be:

```ruby
class ::Topic
  def new_method
    "some response"
  end
end
```

```ruby
add_to_serializer(:topic, :new_method) do
  object.new_method
end
```

Am I right in saying that this kind of thing would now trigger the linting rule, and require me to add an 'ignore' comment?

IMO we should only add rules that are correct in 100% of cases. Otherwise people will just ignore/silence them (as many people have done with the last round of rspec-related rules we enabled)

Reply:
Indeed, this would trigger the rule. At the same time, that’s what we wanted :sweat_smile: (“Methods added to Post, Topic or other classes should check for the enabled site setting”) I agree though that it can quickly become a nuisance, even if that’s an important rule. I tried to find a compromise by only checking `plugin.rb`, but maybe we should remove that rule for now and revisit it later? (or better yet, implement something at the plugin instance level).